cpp/ql/src/Likely Bugs/Underspecified Functions/ImplicitFunctionDeclaration.ql

```diff
@@ -0,0 +1,48 @@
+/**
+ * @name Implicit function declaration
+ * @description An implicitly declared function is assumed to take no
+ * arguments and return an integer. If this assumption does not hold, it
+ * may lead to unpredictable behavior.
+ * @kind problem
+ * @problem.severity warning
+ * @precision medium
+ * @id cpp/implicit-function-declaration
+ * @tags correctness
+ *       maintainability
+ */
+
+import cpp
+import MistypedFunctionArguments
+import TooFewArguments
+import TooManyArguments
+import semmle.code.cpp.commons.Exclusions
+
+predicate locInfo(Locatable e, File file, int line, int col) {
+  e.getFile() = file and
+  e.getLocation().getStartLine() = line and
+  e.getLocation().getStartColumn() = col
+}
+
+predicate sameLocation(FunctionDeclarationEntry fde, FunctionCall fc) {
+  exists(File file, int line, int col |
+    locInfo(fde, file, line, col) and
+    locInfo(fc, file, line, col)
+  )
+}
+
+predicate isCompiledAsC(File f) {
+  f.compiledAsC()
+  or
+  exists(File src | isCompiledAsC(src) | src.getAnIncludedFile() = f)
+}
+
+from FunctionDeclarationEntry fdeIm, FunctionCall fc
+where
+  isCompiledAsC(fdeIm.getFile()) and
+  not isFromMacroDefinition(fc) and
+  fdeIm.isImplicit() and
+  sameLocation(fdeIm, fc) and
+  not mistypedFunctionArguments(fc, _, _) and
+  not tooFewArguments(fc, _) and
+  not tooManyArguments(fc, _)
+select fc, "Function call implicitly declares $@", fc, fc.toString()
```

Comment:
The alert message, when rendered, looks like "Function call implicitly declares `call to igraph_is_simple`", where the bit in monospace is a link to the same location as the main alert location. I think it would be better as "Function call implicitly declares 'igraph_is_simple'", where there's no link. We need links only when the target is different from the location of the main alert location.

A link could be useful for linking to the definition of the called function (if it exists) or to suggest which header file to include. But I suggest leaving that out of this initial PR.

Reply:
Good point. Would it make sense to render it as `Function call implicitly declares 'int igraph_is_simple()'` (as any implicit declaration will receive this signature)?

Reply:
I think that could be confusing. In my experience, most people reading the alert message will think that `int f()` is the same as `int f(void)`. But it should definitely be mentioned and explained in the qhelp.

Reply:
I will keep this in mind as I draft a .qlhelp file, which I think is the next thing on the list for this PR.